From 13410f4fa9fdc533f10495229ccd0e9efc93cc3d Mon Sep 17 00:00:00 2001
From: Marcin Mielczarczyk <marcin.mielczarczyk@redembedded.com>
Date: Fri, 4 Aug 2023 11:04:46 +0200
Subject: [PATCH 08/10] ONEM-30655: Detect Metro subdomain

---
 WebKitBrowser/WebKitImplementation.cpp | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/WebKitBrowser/WebKitImplementation.cpp b/WebKitBrowser/WebKitImplementation.cpp
index c312eb6f..1c3ec0d7 100644
--- a/WebKitBrowser/WebKitImplementation.cpp
+++ b/WebKitBrowser/WebKitImplementation.cpp
@@ -2264,11 +2264,10 @@ static GSourceFuncs _handlerIntervention =
 
             urlValue(URL);
 
-            // TODO: read boot address using DBus!
-            const size_t bootUrl = URL.find("metrological") != string::npos;
-            const size_t blankUrl = URL.find("about:blank") != string::npos;
-            TRACE(Trace::Information, (_T("%s, boot = %d, blank = %d, waitForFailedOrFinished = %d"), URL.c_str(), bootUrl, blankUrl, urlData_.waitForFailedOrFinished));
-            if (bootUrl || blankUrl) {
+            const bool isNewUrlBlankUrl = URL.find("about:blank") != string::npos;
+            static const auto metroDomain = _bootUrl.substr(_bootUrl.find('#'));
+            const bool isNewUrlMetroSubdomain = URL.find(metroDomain) != string::npos;
+            if (isNewUrlBootUrl || isNewUrlBlankUrl || isNewUrlMetroSubdomain) {
                 if (!urlData_.waitForFailedOrFinished) {
                     TRACE(Trace::Information, (_T("Notify that URL has been loaded: %s"), URL.c_str()));
                     std::unique_lock<std::mutex> lock{urlData_.mutex};
-- 
2.25.1

